src/main/java/com/faforever/client/mod/ModUploadTask.java

```diff
@@ -69,17 +85,55 @@ protected Void call() throws Exception {
             .zip();
       }
 
-      log.debug("Uploading mod `{}` as `{}`", modPath, tmpFile);
+      log.debug("Starting upload sequence. Uploading mod `{}` as `{}`", modPath, tmpFile);
       updateTitle(i18n.get("modVault.upload.uploading"));
 
-      return fafApiAccessor.uploadFile("/mods/upload", tmpFile, byteListener, Map.of()).block();
+      return fafApiAccessor.getApiObject(MOD_UPLOAD_START_API_GET, UploadUrlResponse.class)
+                           .flatMap(response -> uploadModToS3(response, tmpFile))
+                           .flatMap(this::completeUpload)
+                           .block();
+
     } finally {
       Files.delete(tmpFile);
       ResourceLocks.freeUploadLock();
     }
   }
 
-  public void setModPath(Path modPath) {
-    this.modPath = modPath;
+  private Mono<Void> completeUpload(UUID requestId) {
+    ModUploadMetadata metadata = new ModUploadMetadata(requestId, null, null);
+    return fafApiAccessor.postJson(MOD_UPLOAD_COMPLETE_API_POST, metadata)
+                         .doOnSuccess(response -> log.debug("Mod upload complete for requestId=[{}]", requestId));
+  }
+
+  private Mono<UUID> uploadModToS3(UploadUrlResponse response, Path filePath) {
+    final URI signedUrl = response.uploadUrl();
+    final UUID requestId = response.requestId();
+    final File modFile = filePath.toFile();
+    final FileSystemResource resource = new FileSystemResource(modFile);
+
+    log.debug("Uploading mod to S3: requestId=[{}], fileName=[{}]", modFile.getName(), requestId);
+
+    return defaultWebClient.put()
+                           .uri(signedUrl)
+                           .accept(MediaType.APPLICATION_JSON)
```

Comment:
Using the defaultWebClient will cause the bearer token to be sent, right?

In my tests with Postman this caused errors on Cloudflare (some signature mismatch or so). Did you test the upload on the test server?

Reply:
No , the `defaultWebClient` bean doesn't have the Exchange Filter Functions. I have tested it with `minio` and the `java-api`
Response from test server:
```
~/test-server$ node server.js
Server running at http://localhost:3000
=== Received PUT Request ===
Headers: {
  host: 'localhost:3000',
  'user-agent': 'downlords-faf-client',
  accept: 'application/json',
  'content-type': 'application/zip',
  'content-length': '64377'
}
```